### DISCUSSION TITLE
Use melos

### DISCUSSION THREAD

**Discussion by fischerscode** [serverpod/serverpod#550] [Ideas]
[Melos](https://melos.invertase.dev/) is a tool that helps managing mono-repos.

I think this repo would benefit from using melos:
- Link packages between each other and run `pub get` using [bootstrap](https://melos.invertase.dev/commands/bootstrap)
- Run `serverpod generate` in every package using [scripts](https://melos.invertase.dev/configuration/scripts)/[exec](https://melos.invertase.dev/commands/exec). (`melos exec -c 1 --depends-on="serverpod" -- "serverpod generate"`)
- Generate changelogs an adjust versions based on commits. [link](https://melos.invertase.dev/commands/version)

**Comment by vlidholt** [comment]
I've been looking at Melos, and it may be a good idea to switch over to it at some point. There were some things (can't remember exactly from the top of my head) that made it non-trivial.

There are some requirements that need to be met to make Serverpod manageable:
- Keeping all the package versions in sync (they should all use the same version).
- Update all READMEs and CHANGELOGs.
- Publish everything, but being able to resume publishing if one package fails to upload.

Something related that I have been thinking of writing a tool for is to check the version of all packages used in the different pubspec files ensure that they are 1. they same version 2. the latest version.

**Reply by fischerscode** [reply]
As far as I know, Melos uses individual versions per package. So maybe it is not the right tool for managing versioning, if you want to keep all versions the same.

But `melos exec` would make the util folder much more elegant.
[generate_all](https://github.com/serverpod/serverpod/blob/main/util/generate_all) would be `melos exec -c 1 --depends-on="serverpod" -- "serverpod generate"`.

**Comment by wstrange** [comment]
Somewhat related, I wanted to use the firebase auth module from git main, but because it has a `path` dependency, pub would not let me use a `git` dependency override in my pubspec.yaml.

If I understand melos, I think it might address this problem by eliminating path dependencies.

**Reply by fischerscode** [reply]
It does, by linking the packages locally using dependency overrides.
